apps/fft/fft.hpp

```diff
@@ -266,93 +274,60 @@ class signal
     std::cout << "]" << std::endl;
   }
 
-  bool isFFT(signal &X, int threads = std::thread::hardware_concurrency())
+  bool isFFT(signal &X, scheduler auto sch, int maxN = 20000)
   {
     int N = y.size();
     bool ret = true;
 
-    data_t *Y = new data_t[N];
-    data_t * M  = new data_t[N*N];
-    auto A = std::mdspan<data_t, view_2d, std::layout_right>(M, N, N);
+    //int nparts = N/maxN;
+    //int psize = std::min(N, nparts);
+    //int matsize = psize * psize;
 
-    // scheduler from a thread pool
-    exec::static_thread_pool ctx{std::min(threads, A.extent(0))};
-    scheduler auto sch = ctx.get_scheduler();
+    std::vector<data_t> Y(N);
+    std::vector<data_t> M(N*N);
 
-    ex::sender auto test = ex::bulk(schedule(sch), A.extent(0), [&](int i){
-      for (auto j = 0; j < A.extent(1); j++){
-        A(i, j) = WNk(N, i*j);
-      }
-    })
-    // Compute fft
-    | ex::bulk(A.extent(0), [&](int i){
-      for (auto j = 0; j < A.extent(1); j++){
-        Y[i]+= A(i,j) * y[j];
-      }
-    })
-    // compare the computed fft with input
-    | ex::bulk(N, [&](int i){
-      if (!complex_compare(X[i], Y[i]))
+    auto A   = std::mdspan<data_t, view_2d, std::layout_right>(M.data(), N, N);
+    auto mdy = std::mdspan<data_t, view_2d, std::layout_right>(y.data(), N, 1);
+    auto mdY = std::mdspan<data_t, view_2d, std::layout_right>(Y.data(), N, 1);
+
+    data_t *F = M.data();
+    data_t *X_ptr = X.data();
+    data_t *Y_ptr = Y.data();
+
+    ex::sender auto init = ex::transfer_just(sch, F) | ex::bulk(N*N, [=](int k, auto F){
+      int i = k / N;
+      int j = k % N;
+      F[k] = WNk(N, i*j);
+    });
+
+    // initialize
+    ex::sync_wait(init);
+
+    // compute Y[n] = dft(x[n]) = WNk * x[n]
+    stdex::linalg::matrix_product(std::execution::par, A, mdy, mdY);
```

Comment:
I remembered to use stdpar=gpu, we need `par_unseq` execution policy, but I did not find any performance gain at some point. Maybe we can ask Nvidia folks about it.